### ISSUE TITLE
 Python 3.12 support

### ISSUE THREAD

**Issue by juhoinkinen** [NatLibFi/Annif#779]
Python 3.12 was [released last October](https://www.python.org/downloads/release/python-3120/), so we should try to support it in Annif.

Currently support is not possible because of the following (to be continued...):
- ~gensim v4.3.2 does not provide a wheel for Python 3.12 in [PyPI](https://pypi.org/project/gensim/#files), and building the wheel locally fails:~

    ```
    ...
    gensim/models/word2vec_inner.c:25:10: fatal error: Python.h: No such file or directory
         25 | #include "Python.h"
            |          ^~~~~~~~~~
      compilation terminated.
      error: command '/usr/bin/x86_64-linux-gnu-gcc' failed with exit code 1
    ```
   - ~issues https://github.com/piskvorky/gensim/issues/3511, https://github.com/piskvorky/gensim/issues/3510~
- tensorflow-io-gcs-filesystem requires Python >=3.7, <3.12 
  - issues https://github.com/tensorflow/io/issues/1761, https://github.com/tensorflow/io/issues/1966)
- stwfsapy requires Python <3.12,>=3.9 
    - issue https://github.com/zbw/stwfsapy/issues/61

**Comment by juhoinkinen** [comment]
> - gensim v4.3.2 does not provide a wheel for Python 3.12 in [PyPI](https://pypi.org/project/gensim/#files), and building > the wheel locally fails:
>
 >   ```
>    ...
>    gensim/models/word2vec_inner.c:25:10: fatal error: Python.h: No such file or directory
>         25 | #include "Python.h"
>            |          ^~~~~~~~~~
>      compilation terminated.
>      error: command '/usr/bin/x86_64-linux-gnu-gcc' failed with exit code 1
>    ```
>   - issues https://github.com/piskvorky/gensim/issues/3511, https://github.com/piskvorky/gensim/issues/3510

Actually I got this resolved by installing python3.12-dev system package.

**Comment by juhoinkinen** [comment]
As gensim can be installed, provided that the `python3.12-dev` system package is available, even currently it is possible to install Annif (or at least Poetry can resolve dependencies) if the unsupported dependencies are resctricted to Python <3.12, that is using this in `pyproject.toml`:
```
tensorflow-cpu = { version = "2.15.*", optional = true, python = "<3.12" }
```
```
stwfsapy = { version = "0.4.*", optional = true, python = "<3.12" }
```

**Comment by osma** [comment]
It looks like Ubuntu 24.04, [to be released in April 2024](https://wiki.ubuntu.com/Releases), will [come with Python 3.12](https://launchpad.net/ubuntu/noble/+package/python3-all). So it would be nice if we could support it soon. But of course, older Python versions can also be installed on Ubuntu.

**Comment by juhoinkinen** [comment]
> ```
> tensorflow-cpu = { version = "2.15.*", optional = true, python = "<3.12" }
> ```

Also TensorFlow could be installed, if it could be upgraded to [version 2.16.*](https://github.com/tensorflow/tensorflow/releases/tag/v2.16.1), but this is blocked by https://github.com/tensorflow/tensorflow/issues/63585.

**Comment by juhoinkinen** [comment]
Not fully closed, because stwfsapy and NN ensemble do not work on Python 3.12.

**Comment by juhoinkinen** [comment]
Closed by #796.
